racker/salus-telemetry-envoy
# Set net_raw capabilities on installed telegraf binary and tighten up install locaitons

# Resolves

https://jira.rax.io/browse/SALUS-595

# What

This Jira issue expanded beyond what I initially intended after looking at how [telegraf's packaging](https://github.com/influxdata/telegraf/tree/master/scripts) did the non-root user setup. From that I noticed specifically that the install location for systemd files differs between the "non standard" location used by Ubuntu/Debian and everything else, such as RHEL/CentOS

https://github.com/influxdata/telegraf/blob/master/scripts/post-install.sh#L50

Back to the original issue, the original point of these changes was to ensure that the telegraf binary had the `net_raw` capability set so that it can have the option of using native ping operations. BTW, I discovered this change isn't even needed prior to telegraf 1.12 since it previously used the system's `ping` command anyway.

# How

* Brought in a module that is able to make system calls to set capabilities on files. I originally tried using `setcap` system command, but the separation of process scopes defeated the `setfcap` capability we have set on Envoy.
* Added a `PostInstall` hook to the per-agent code to let the telegraf one set the net_raw capability with that new module
* Updated install scripts to create a user for `telemetry-envoy`
* Updated systemd config to run as that user and conditionally start when main config file is present
* Install the template config file separately with a `.tmpl` suffix to make the service conditional logic more apparent
* Introduced post install/remove scripts to handle file ownership, portable systemd service file locations, etc

## How to test

The capabilities aspect was tested manually using the new `detached` sub-command and providing this instructions file:

```json
{
  "instructions": [
    {
      "install": {
        "agent": {
          "version": "1.12.0"
        },
        "url": "https://dl.influxdata.com/telegraf/releases/telegraf-1.12.0-static_linux_amd64.tar.gz",
        "exe": "./telegraf/telegraf"
      }
    },
    {
      "configure": {
        "operations": [
          {
            "id": "monitor-1",
            "content": "{\"type\": \"ping\",\"urls\":[\"rackspace.com\"],\"method\":\"native\"}",
            "extraLabels": {
              "detached": "true"
            },
            "interval": "30"
          }
        ]
      }
    }
  ]
}
```

Packaging changes were manually tested with the following test sequence on Ubuntu cloud server:

1. Install package
2. Confirm service is present, but inactive with `systemctl status telemetry-envoy`
3. Populate `/etc/salus/telemetry-envoy.yaml` making sure it is owned by `telemetry-envoy` user. Can use `eletest` user and API key for this.
4. Start the service with `systemctl start telemetry-envoy`
5. Confirm started using `systemctl status telemetry-envoy`
6. Perform an upgrade, such as by bumping the `epoch` in the `nfpms` section of `.goreleaser.yml`
7. Confirm service was restarted and is still active
8. Confirm new package version was installed using `dpkg --status telemetry-envoy`. In the `Version`, the left-most digit is the epoch, such as "1" in `Version: 1:0.11.0`
9. Perform a package remove (not purge)
10. Confirm service is stopped using `journalctl -u telemetry-envoy` and unit is removed using `systemctl status telemetry-envoy`
11. Confirm `/etc/salus` still has the modified file, but template file in there got removed along with `/usr/local/bin/telemetry-salus`
12. On Ubuntu/Debian, purge the package
13. Confirm `/etc/salus` and `/var/lib/telemetry-envoy` has been removed 

Tested an abbreviated version on a CentOS cloud server.
